core/src/main/java/jenkins/util/URLClassLoader2.java

```diff
@@ -0,0 +1,49 @@
+package jenkins.util;
+
+import jenkins.ClassLoaderReflectionToolkit;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import java.net.URL;
+import java.net.URLClassLoader;
+
+/**
+ * {@link URLClassLoader} with loosened visibility for use with {@link
+ * ClassLoaderReflectionToolkit}.
+ */
+@Restricted(NoExternalUse.class)
+public class URLClassLoader2 extends URLClassLoader implements JenkinsClassLoader {
+
+    static {
+        registerAsParallelCapable();
+    }
+
+    public URLClassLoader2(URL[] urls) {
+        super(urls);
+    }
+
+    public URLClassLoader2(URL[] urls, ClassLoader parent) {
+        super(urls, parent);
+    }
+
+    @Override
+    public void addURL(URL url) {
+        super.addURL(url);
+    }
```

Comment:
Why not just define this in `JenkinsClassLoader`?

Reply:
You are adding a public `addURL` method in this PR, with the intention of allowing a revision of https://github.com/jenkinsci/bouncycastle-api-plugin/pull/46 to call it. You are also defining a `JenkinsClassLoader` to abstract the differences between `AntClassLoader` and `URLClassLoader2`. Rather than having that plugin have to go through a list of implementation classes looking for a method of this kind it could call, I am suggesting that the method be defined in one place, `JenkinsClassLoader`, with `AntClassLoader.addPathComponent` made into an implementation. So it seems in line with the goals of this PR; unless you want to _not_ do anything to help `bouncycastle-api` here, in which case there should be no need for a public `addURL` override either.

Reply:
(This is a suggestion, not a blocking comment; I have already approved this PR.)

Reply:
> Today, the BouncyCastle API plugin calls a public method on a restricted AntClassLoader class, and my intention is to allow it to also call a public method on a restricted URLClassLoader2 class. This preserves the status quo, making things neither better nor worse.

Actually as I understand this, and this is my concern with the PR - it does make it slightly worse for users.
this PR as-is is not useful for the vast majority of users (if I read the Jenkins stats correctly almost every instance of Jenkins has the bouncycastle plugin installed - thus getting this confidence would take a long time) and by the time we enable it they would also have to have upgraded the BC plugin prior to that (so that is a lead time of 2 LTS bumps ~6 months) (and at least 3 months before users can opt in to test this _without branching the BC plugin_)
now bouncycastle is a detached plugin so we could update the version that is in the war, but if I recall correctly they do not get updated if already installed.

Would it not be better to put the code from `URLClassloader2` into the ANTClassloader, and the class that currently extends it can then take the code that is currently in `ANTClassloader`  that way - there is no need for a switch, the bouncycastle plugin continues to work, users do not need to ensure they have a compatible BC plugin when they upgrade a core that has the flag defaulted to true...

If we are not confident that we can use a subclass of `URLClassLoader` then what would make that possible?
If we are confident then the different approach would also seem to work as it is not changing the class hierarchy and should be binary compatible for everyone.

Happy to do an alternative PR to sketch this approach out.

Reply:
Not sure what this has to do with LTS?

Can't Bouncycastle API update to the weekly this is released in, and then add support for the `URLClassloader2` and release a new version?

and in the upgrade note when we get to LTS specify to make sure the plugin is upgraded?

Reply:
well a user on an LTS can not upgrade BC to a version with the changes, so you need to wait 2 LTS versions until you pull the plug (~ 5 months from now at the earliest). Basically LTS adds significant delays to be able to actually enable this flag (stats seems to imply > 90% of Jenkins installs have BC plugin installed) because you can not change the default until a user can install a version of the plugin that is compatible with core **before** they upgrade to that core (using the UI to manage plugin upgrades).

Sure we can branch the BC plugin - but how much testing would the weekly have? And if we think the URLClassloader is fine (I think we all do) then why do we even need to have something that breaks compatibility when I believe there is an approach that works regardless and requires no changes?

Reply:
I disagree about all this LTS talk, you don't need to wait, backports can be easily made if required, and BC doesn't appear to release very often anyway: https://mvnrepository.com/artifact/org.bouncycastle/bcprov-jdk15on.

If you have a suggestion and are happy to provide an alternate PR then of course that will be looked at.

Reply:
We could also revert this change to bouncy castle that took this internal implementation detail and get a supported public API added first?

Reply:
> you don't need to wait, backports can be easily made if required

I think you have completely missed my point - you can not upgrade Jenkins and BC atomically.
I could release a BC targetting tomorrows weekly sure and in the next weekly we flip the flag.

How would an LTS user upgrade to the next .1 release if they have BC plugin installed without having a broken jenkins instance for some period of time?

Reply:
`bouncycastle-api` is already using pretty awful hacks; at worst we could cut a release that had an older `jenkins.version` but worked with various implementation classes via reflection, then promptly push a new release with the new `jenkins.version` without the reflection.

(But still awful hacks. The root problem here IIUC is that BC itself is not designed correctly—you should be able to swap in a FIPS or standard implementation without replacing the API types.)

Reply:
you can not use BC-FIPS inside a plugin period.  the JVM needs to be configured and doing that in a plugin is too late as the JVM is up and configured, jetty has started already initialized some SSL context etc etc.

The API is mostly compatable (if you stay away from deprecated types).  not all providers are available in the regular BC (yup - there are some bits only in FIPS, when you expect the other way around).

we have been mostly able to swap a FIPS and non FIPS versions which is how the plugin can still work, but it is not possible to install a FIPS security provider in anything other than the java security config (or at best the first line of code in the `main` method.

Reply:
> We could also revert this change to bouncy castle that took this internal implementation detail and get a supported public API added first?

so if the BouncyCastle API is going to use reflection to be compatible with this new feature I would not be overly happy doing that unless the new feature to be used (via reflection) was a public (aka not restricted) API - otherwise we will be in the same boat we are in today, but with reflection thrown into the mix for added fun.

Currently this is not compatible and adds no public API that can be migrated to.

It can not be reverted as it is actually required, but introducing reflection with a supported API is a path I would support if we are not going to go with an approach that maintains the compatability.